crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -445,17 +449,96 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    let upgraded_package_modules = publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Read the package dependencies.
+    let dependency_packages = dep_ids
+        .iter()
+        .map(|id| fetch_package(context, id))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // TODO(tzakian): check digest of these packages against the digest that we expect in the
+    // upgrade ticket. Do we need to check that we don't have the same package at different
+    // versions in the depencency list?
+
+    // Full backwards compatibility except that we allow friend function signatures to change.
+    let compatiblity_checker = Compatibility::new(true, true, false);
+
+    // Run compatibility checks against the old package that was there.
+    // NB: due to mutability reasons and to minimize clones we want to separate the
+    // `upgrade_package` and `add_package` steps so `context` can be an immutable borrow. This is
+    // important since the references to both `current_package` and all the packages in
+    // `dependency_packages` are rooted in `context`.
+    let upgraded_package_object = context.upgrade_package(
+        current_package,
+        upgraded_package_modules,
+        dependency_packages,
+        compatiblity_checker,
+    )?;
+    let upgraded_object_id = upgraded_package_object.id();
+    context.add_package(upgraded_package_object);
+
+    Ok(vec![Value::Object(ObjectValue::new(
+        UpgradeReceipt::type_().into(),
+        /* has_public_transfer */ true,
+        /* used_in_non_entry_move_call */ false,
+        &bcs::to_bytes(&UpgradeReceipt::new(upgrade_ticket, upgraded_object_id)).unwrap(),
+    )?)])
+}
+
+fn fetch_package<'a, E: fmt::Debug, S: StorageView<E>>(
+    context: &'a ExecutionContext<E, S>,
+    package_id: &ObjectID,
+) -> Result<&'a MovePackage, ExecutionError> {
+    let object = context.state_view.read_object(package_id).ok_or_else(|| {
```

Comment:
Transferring the question from the previous PR -- `state_view` is a `TemporaryStore` -- will it be able to read objects that aren't in the input objects of the transaction? cc @tnowacki

Reply:
Talked to @tzakian offline, should be good here